packages/admin/cms-admin/src/generator/future/generateGrid.ts

```diff
@@ -57,6 +58,164 @@ function findInputObjectType(input: IntrospectionInputValue, schema: Introspecti
+                    }`,
+                );
+            } else {
+                requiredParams.push({ type: arg.type.ofType.name, name: arg.name });
+            }
+        } else if (arg.type.ofType.kind === "INPUT_OBJECT") {
+            requiredParams.push({ type: arg.type.ofType.name, name: arg.name });
+        } else if (arg.type.ofType.kind === "LIST") {
+            console.warn(
+                `Currently not supported kind LIST in required param ${arg.name} of ${gridQueryType.name}${
+                    createMutationType ? ` or ${createMutationType.name}` : ``
+                }`,
+            );
+        } else if (arg.type.ofType.kind === "ENUM") {
+            console.warn(
+                `Currently not  supported kind ENUM in required param ${arg.name} of ${gridQueryType.name}${
+                    createMutationType ? ` or ${createMutationType.name}` : ``
+                }`,
+            );
+        } else {
+            throw new Error(`Not supported kind ${arg.type.ofType.kind}`);
+        }
+    });
+    return requiredParams;
+}
+
+function getRequiredGqlArgTypesForImport({
+    config,
+    gridQueryType,
+    createMutationType,
+    gqlIntrospection,
+}: {
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    config: GridConfig<any>;
+    gridQueryType: IntrospectionField;
+    createMutationType?: IntrospectionField;
+    gqlIntrospection: IntrospectionQuery;
+}) {
+    const types: string[] = [];
+    getRequiredGqlArgs({ config, gridQueryType, createMutationType, gqlIntrospection }).forEach((arg) => {
+        const nativeScalars = ["ID", "String", "Boolean", "Int", "Float", "DateTime", "JSONObject"];
+        if (nativeScalars.includes(arg.type)) return;
+        types.push(arg.type);
+    });
+    return types;
+}
+
+function generateGridPropsType({
+    config,
+    gridQueryType,
+    createMutationType,
+    gqlIntrospection,
+}: {
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    config: GridConfig<any>;
+    gridQueryType: IntrospectionField;
+    createMutationType?: IntrospectionField;
+    gqlIntrospection: IntrospectionQuery;
+}) {
+    const props: string[] = [];
+    getRequiredGqlArgs({ config, gridQueryType, createMutationType, gqlIntrospection }).forEach((arg) => {
+        let type = arg.type;
+        if (arg.type === "ID" || arg.type === "String" || arg.type === "DateTime") {
+            type = "string";
+        } else if (arg.type === "Boolean") {
+            type = "boolean";
+        } else if (arg.type === "Int" || arg.type === "Float") {
+            type = "number";
+        } else if (arg.type === "JSONObject") {
+            type = "unknown"; // because any needs eslint-disable
+        }
+        props.push(`${arg.name}: ${type};`);
+    });
+    return props.length
+        ? `type Props = {
+        ${props.join("\n")}
+    };`
+        : undefined;
+}
+
+function generateGridProps({
```

Comment:
Could you please move generateGridProps, generateGridPropsType, getRequiredGqlArgTypesForImport and getRequiredGqlArgs into a single function that returns everything you need for the 4 use cases?

That would even allow you to move getRequiredQueryArgs and getRequiredMutationArgs as simple variable into that function.

I think that will make this much easier to understand. One function for one concern.

Reply:
I think this should be addressed in a special refactoring pull-request. Because I also see the generation of different code-parts as a concern of their own, because they need to know/merge the different features to some degree. (you can see this for example in the GridPropsType generation. GridPropsType needs to be defined if there are required gql-vars **or** if filter-props should be exposed #1862. so both features/concerns need to be checked to find out if the surrounding `type Props = { ... }` has to be rendered, because prettier-fix does not remove an empty type.)

I already got some rough idea how we can solve this issue. There are some similarities to my initial admin-generator draft, but hopefully less overengineered :-D